ray-on-gke/user/kuberay-tpu-webhook/main.go

```diff
@@ -0,0 +1,294 @@
+	pod, err := extractPod(admissionReview)
+	if err != nil {
+		klog.Fatalf("Pod extraction failed: %s", err)
+	}
+
+	var patches []patch
+	// ray operator only sets GenerateName field - doesn't include random suffix until after admission request
+	// use mapping of {cluster name, group name} -> # workers created to set TPU_WORKER_IDs
+	clusterName := pod.Labels["ray.io/cluster"]
+	groupName := pod.Labels["ray.io/group"]
+	podSlice := slice{clusterName, groupName}
+	topology := pod.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+	isMultiHost, err := isTPUMultiHost(topology)
+	if err != nil {
+		return nil, err
+	}
+
+	// inject the TPU_WORKER_ID environment variable into the container requesting TPUs
+	if containerRequestingTPUs(pod.Spec.Containers...) && isMultiHost {
+		// assign to the next unique ID in the pod slice
+		tpu_worker_id := sliceToWorkers[podSlice]
+		sliceToWorkers[podSlice] += 1
+		for i := 0; i < len(pod.Spec.Containers); i++ {
+			container := pod.Spec.Containers[i]
+			if containerRequestingTPUs(container) {
+				path := fmt.Sprintf("/spec/containers/%d/env", i)
+				tpuWorkerID := corev1.EnvVar{
+					Name:  "TPU_WORKER_ID",
+					Value: fmt.Sprint(tpu_worker_id),
+				}
+				tpuName := corev1.EnvVar{
+					Name:  "TPU_NAME",
+					Value: fmt.Sprint(groupName),
+				}
+				idPatch, namePatch := patch{"op": "add",}, patch{"op": "add",}
+				if len(container.Env) == 0 {
+					idPatch["path"], namePatch["path"] = path, path
+					idPatch["value"] = []corev1.EnvVar{tpuWorkerID}
+					namePatch["value"] = []corev1.EnvVar{tpuName}
+				} else {
+					idPatch["path"], namePatch["path"] = fmt.Sprintf("%s/-", path), fmt.Sprintf("%s/-", path)
+					idPatch["value"] = tpuWorkerID
+					namePatch["value"] = tpuName
+				}
+				patches = append(patches, idPatch, namePatch)
+			}
+		}
+	}
+
+	patchBytes, err := json.Marshal(patches)
+	if err != nil {
+		klog.Fatalf("Error serializing patches: %s", err)
+	}
+
+	admissionResponse := &admissionv1.AdmissionResponse{
+		UID: 	 admissionReview.Request.UID,
+		Allowed: true,
+		Patch:   patchBytes,
+		PatchType: func() *admissionv1.PatchType {
+			pt := admissionv1.PatchTypeJSONPatch
+			return &pt
+		}(),
+	}
+	return admissionResponse, nil
+}
+
+func init() {
+	sliceToWorkers = make(map[slice]int)
+}
+
+func main() {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprint(w, "kuberay-tpu-webhook")
+	})
+	mux.HandleFunc("/inject", func(w http.ResponseWriter, r *http.Request) {
+		admissionReview := &admissionv1.AdmissionReview{}
+		if err := json.NewDecoder(r.Body).Decode(admissionReview); err != nil {
+			http.Error(w, "Error decoding request body", http.StatusBadRequest)
+			return
+		}
+
+		if admissionReview.Request.Kind.Kind == "RayCluster" {
+			klog.Info("Received review for RayCluster")
+			admissionReview.Response, _ = mutateRayCluster(admissionReview)
```

Comment:
I see there's 4 instances of ignoring returned errors in this main() function, is there a reason we're doing that?  I think this could lead to unexpected behavior. I would suggest logging the error and returning early unless there is a reason we need to ignore the errors.

Reply:
Done.